README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->
## scrolling_text

 This package can be used to make a scrolling text widget(marquee) same as shown on bottom of the news channel.

## Features
- custom text style
- start and ending offset for text
- you can have callback function on the end of any text scrolling 
- controlled speed with speed and rest parameters
- vertical or horizontal scroll direction



## Usage
Add this Project to Pubspec and import as

import 'package:scrolling_text/scrolling_text.dart';

You can use as this to play multiple text in marquee

```dart
class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Scrolling Text Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: Scaffold(
          body: Center(
            child: Container(
              width: 30,
              color: const Color.fromARGB(255, 128, 184, 230),
              padding: const EdgeInsets.all(10),
              child: ScrollingText(
                text: news[index],
                scrollAxis: Axis.vertical,
                onFinish: () {
                  if (index < news.length) {
                    index++;
                  } else {
                    index = 0;
                  }
                  setState(() {});
                },
              ),
            ),
          ),
        ));
  }
}

int index = 0;
List<String> news = [
  'Lorem ipsum is a placeholder text commonly used to demonstrate the visual form of a document or a typeface without relying on meaningful content',
  'Lorem Ipsum is simply dummy text of the printing and typesetting industry',
  'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using  making it look like readable English.'
];

```
## Basic Results

Horizontal Example:

![](https://github.com/innovatoraakash/text_marquee_package/blob/master/assets/horizontal.gif)

Vertical Example

![](https://github.com/innovatoraakash/text_marquee_package/blob/master/assets/vertical.gif)



CHANGELOG.md
## 0.0.4

* TODO: Proper Formating.


test/text_marquee_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:text_marquee/text_marquee.dart';

void main() {

}


test/scrolling_text_test.dart
import 'package:flutter/material.dart';

import 'package:scrolling_text/scrolling_text.dart';

void main() {}

class TestScreen extends StatefulWidget {
  const TestScreen({super.key});

  @override
  State<TestScreen> createState() => _TestScreenState();
}

class _TestScreenState extends State<TestScreen> {
  @override
  Widget build(BuildContext context) {
    return ScrollingText(
      text: "",
      onFinish: () {},
    );
  }
}


lib/text_marquee.dart
library text_marquee;

import 'dart:async';

import 'package:flutter/material.dart';



class ScrollingText extends StatefulWidget {

  final String text;
  final TextStyle? textStyle;
  final Axis scrollAxis;
  final double ratioOfBlankToScreen;
  final double speed;
  final int restTime;
  final VoidCallback onFinish;

 const ScrollingText(
      {super.key,
        required this.text,
      this.textStyle,
      this.scrollAxis= Axis.horizontal,
      this.ratioOfBlankToScreen= 0.9,
      this.speed= 60.0,
      this.restTime= 1000,
      required this.onFinish})
      ;

  @override
  State<StatefulWidget> createState() {
    return ScrollingTextState();
  }
}

class ScrollingTextState extends State<ScrollingText>
    with SingleTickerProviderStateMixin {
  ScrollController? scrollController;
  double? screenWidth;
  double? screenHeight;
  double position = 0.0;
  Timer? timer;
  GlobalKey _key = GlobalKey();

  @override
  void initState() {
    super.initState();
    scrollController = ScrollController();
    WidgetsBinding.instance.addPostFrameCallback((callback) {
      startTimer();
    });
  }

  void startTimer() {
    if (_key.currentContext != null) {
      double widgetWidth =
          _key.currentContext!.findRenderObject()!.paintBounds.size.width;
      double widgetHeight =
          _key.currentContext!.findRenderObject()!.paintBounds.size.height;

      timer = Timer.periodic(Duration(milliseconds: widget.restTime), (timer) {
        double maxScrollExtent = scrollController!.position.maxScrollExtent;
        double pixels = scrollController!.position.pixels;
        if (pixels + widget.speed >= maxScrollExtent) {
          position = 0.0;
          scrollController!.jumpTo(position);
          widget.onFinish();
        }
        position += widget.speed;
        scrollController!.animateTo(position,
            duration: Duration(milliseconds: widget.restTime), curve: Curves.linear);
      });
    }
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    screenWidth = MediaQuery.of(context).size.width;
    screenHeight = MediaQuery.of(context).size.height;
  }

  Widget getBothEndsChild() {
    if (widget.scrollAxis == Axis.vertical) {
      String newString = widget.text.split("").join("\n");
      return Text(
        newString,
        style: widget.textStyle,
        textAlign: TextAlign.right,
      );
    }
    return Center(
      child: Text(
        widget.text,
        style: widget.textStyle,
      ),
    );
  }

  Widget getCenterChild() {
    return Container(width: screenWidth! * widget.ratioOfBlankToScreen);
  }

  @override
  void dispose() {
    super.dispose();
    if (timer != null) {
      timer!.cancel();
    }
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: ListView(
        key: _key,
        scrollDirection: widget.scrollAxis,
        controller: scrollController,
        physics: NeverScrollableScrollPhysics(),
        children: <Widget>[
          getCenterChild(),
          getBothEndsChild(),
          getCenterChild(),
        ],
      ),
    );
  }
}



lib/scrolling_text.dart
library scrolling_text;

import 'dart:async';

import 'package:flutter/material.dart';

class ScrollingText extends StatefulWidget {
  ///[text] is a required field which is animated in vertical/ horizontal direction
  final String text;

  ///[TextStyle] class support styling of the [text] parameter
  final TextStyle? textStyle;

  /// defines scroll direction of the text.
  ///  [Axis.horizontal] is a default
  final Axis scrollAxis;

  /// speed of [text] to scroll.
  ///
  /// default speed is 60 pixels per [restTime].
  /// relative to [restTime]
  final double speed;

  ///Starting blank offset to the content i.e [text]
  ///
  final double? startOffset;

  ///ending blank offset to the content i.e [text]
  ///
  final double? endOffset;

  ///It defines speed of [text] to scroll.
  ///
  /// default [restTime] is 1000 milliseconds
  /// relative to [speed]
  final int restTime;

  /// [Curve] class defines way how text scroll inside the widget
  ///
  /// [Curves.linear] is a default [scrollCurve]
  final Curve? scrollCurve;

  /// this is called whenever a [text] mequeue finishes to scroll along with [startOffset] and [endOffset]
  ///
  /// specially call this to change the content from outside the class

  final VoidCallback onFinish;

  ///Scorlling Text provides marquee of text with callback at the end of scroll and also staring and ending offset of the text

  const ScrollingText(
      {super.key,
      required this.text,
      this.textStyle,
      this.scrollAxis = Axis.horizontal,
      this.speed = 60.0,
      this.startOffset,
      this.endOffset,
      this.restTime = 1000,
      this.scrollCurve,
      required this.onFinish});

  @override
  State<StatefulWidget> createState() {
    return ScrollingTextState();
  }
}

class ScrollingTextState extends State<ScrollingText>
    with SingleTickerProviderStateMixin {
  ScrollController? scrollController;
  double? screenWidth;
  double? screenHeight;
  double position = 0.0;
  Timer? timer;
  GlobalKey key = GlobalKey();

  @override
  void initState() {
    super.initState();
    scrollController = ScrollController();
    WidgetsBinding.instance.addPostFrameCallback((callback) {
      startTimer();
    });
  }

  void startTimer() {
    if (key.currentContext != null) {
      // double widgetWidth =
      //     _key.currentContext!.findRenderObject()!.paintBounds.size.width;
      // double widgetHeight =
      //     _key.currentContext!.findRenderObject()!.paintBounds.size.height;

      timer = Timer.periodic(Duration(milliseconds: widget.restTime), (timer) {
        double maxScrollExtent = scrollController!.position.maxScrollExtent;
        double pixels = scrollController!.position.pixels;
        if (pixels + widget.speed >= maxScrollExtent) {
          position = 0.0;
          scrollController!.jumpTo(position);
          widget.onFinish();
        }
        position += widget.speed;
        scrollController!.animateTo(position,
            duration: Duration(milliseconds: widget.restTime),
            curve: widget.scrollCurve ?? Curves.linear);
      });
    }
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    screenWidth = MediaQuery.of(context).size.width;
    screenHeight = MediaQuery.of(context).size.height;
  }

  Widget getBothEndsChild() {
    if (widget.scrollAxis == Axis.vertical) {
      String newString = widget.text.split("").join("\n");
      return Text(
        newString,
        style: widget.textStyle,
        textAlign: TextAlign.right,
      );
    }
    return Center(
      child: Text(
        widget.text,
        style: widget.textStyle,
      ),
    );
  }

  Widget getCenterChild(bool isStart) {
    if (widget.scrollAxis == Axis.vertical) {
      return Container(height: isStart ? widget.startOffset : widget.endOffset);
    }
    return Container(width: isStart ? widget.startOffset : widget.endOffset);
  }

  @override
  void dispose() {
    super.dispose();
    if (timer != null) {
      timer!.cancel();
    }
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: ListView(
        key: key,
        scrollDirection: widget.scrollAxis,
        controller: scrollController,
        physics: const NeverScrollableScrollPhysics(),
        children: <Widget>[
          getCenterChild(true),
          getBothEndsChild(),
          getCenterChild(false),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:scrolling_text/scrolling_text.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Scrolling Text Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: Scaffold(
          body: Center(
            child: Container(
              width: 30,
              color: const Color.fromARGB(255, 128, 184, 230),
              padding: const EdgeInsets.all(10),
              child: ScrollingText(
                text: news[index],
                scrollAxis: Axis.vertical,
                onFinish: () {
                  if (index < news.length) {
                    index++;
                  } else {
                    index = 0;
                  }
                  setState(() {});
                },
              ),
            ),
          ),
        ));
  }
}

int index = 0;
List<String> news = [
  'Lorem ipsum is a placeholder text commonly used to demonstrate the visual form of a document or a typeface without relying on meaningful content',
  'Lorem Ipsum is simply dummy text of the printing and typesetting industry',
  'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using  making it look like readable English.'
];
